beater/beater.go

```diff
@@ -310,63 +332,75 @@ type sharedServerRunnerParams struct {
 }
 
 func newServerRunner(ctx context.Context, args serverRunnerParams) (*serverRunner, error) {
-	cfg, err := config.NewConfig(args.RawConfig, elasticsearchOutputConfig(args.Beat))
+	runServerContext, cancel := context.WithCancel(ctx)
+	config, err := config.NewConfig(args.RawConfig, elasticsearchOutputConfig(args.Beat))
 	if err != nil {
 		return nil, err
 	}
-
-	runServerContext, cancel := context.WithCancel(ctx)
-	return &serverRunner{
+	s := &serverRunner{
 		backgroundContext:      ctx,
 		runServerContext:       runServerContext,
 		cancelRunServerContext: cancel,
-
-		config:        cfg,
-		rawConfig:     args.RawConfig,
-		fleetConfig:   args.FleetConfig,
-		acker:         args.Acker,
-		pipeline:      args.Pipeline,
-		namespace:     args.Namespace,
-		beat:          args.Beat,
-		logger:        args.Logger,
-		tracer:        args.Tracer,
-		tracerServer:  args.TracerServer,
-		wrapRunServer: args.WrapRunServer,
-	}, nil
-}
-
-func (s *serverRunner) String() string {
-	return "APMServer"
+		config:                 config,
+		rawConfig:              args.RawConfig,
+		fleetConfig:            args.FleetConfig,
+		acker:                  args.Acker,
+		pipeline:               args.Pipeline,
+		namespace:              args.Namespace,
+		beat:                   args.Beat,
+		logger:                 args.Logger,
+		tracer:                 args.Tracer,
+		tracerServer:           args.TracerServer,
+		wrapRunServer:          args.WrapRunServer,
+		restartc:               make(chan struct{}),
+		// This is just for testing right now.
+		params: args,
+	}
+	if err := s.configure(args); err != nil {
+		return nil, err
+	}
+	return s, nil
 }
 
-// Stop stops the server.
-func (s *serverRunner) Stop() {
-	s.stopOnce.Do(s.cancelRunServerContext)
-	s.Wait()
-}
+func (s *serverRunner) configure(args serverRunnerParams) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
 
-// Wait waits for the server to stop.
-func (s *serverRunner) Wait() {
-	s.wg.Wait()
-}
+	config, err := config.NewConfig(args.RawConfig, elasticsearchOutputConfig(args.Beat))
+	if err != nil {
+		return err
+	}
 
-// Start starts the server.
-func (s *serverRunner) Start() {
-	s.wg.Add(1)
-	go func() {
-		defer s.wg.Done()
-		s.run()
-	}()
-}
+	// once the server has been created, we only want to update the values
+	// pointed at, not the pointer itself.
+	*s.config = *config
+	if args.RawConfig != nil {
+		*s.rawConfig = *args.RawConfig
+	}
+	if args.FleetConfig != nil {
+		*s.fleetConfig = *args.FleetConfig
+	}
+	if args.Acker != nil {
+		*s.acker = *args.Acker
+	}
+	s.pipeline = args.Pipeline
+	s.namespace = args.Namespace
+	if args.Beat != nil {
+		*s.beat = *args.Beat
+	}
+	if args.Logger != nil {
+		*s.logger = *args.Logger
+	}
+	if args.Tracer != nil {
+		*s.tracer = *args.Tracer
+	}
+	if args.TracerServer != nil {
+		*s.tracerServer = *args.TracerServer
+	}
+	s.wrapRunServer = args.WrapRunServer
```

Comment:
The runServer concept (I expect) still needs to be dealt with -- as @axw mentioned, we might need to have a pre-set chained BatchProcessor and make them configurable as the conf file changes.